pennylane/devices/reference_qubit.py

```diff
@@ -0,0 +1,152 @@
+
+
+def supports_operation(op: qml.operation.Operator) -> bool:
+    """This function used by preprocessing determines what operations
+    are natively supported by the device.
+
+    While in theory ``simulate`` can support any operation with a matrix, we limit the target
+    gate set for improved testing and reference purposes.
+
+    """
+    return getattr(op, "name", None) in operations
+
+
+@simulator_tracking  # update device.tracker with some relevant information
+@single_tape_support  # add support for device.execute(tape) in addition to device.execute((tape,))
+class ReferenceQubit(Device):
+    """A slimmed down numpy-based simulator for reference and testing purposes.
+
+    Args:
+        wires (int, Iterable[Number, str]): Number of wires present on the device, or iterable that
+            contains unique labels for the wires as numbers (i.e., ``[-1, 0, 2]``) or strings
+            (``['ancilla', 'q1', 'q2']``). Default ``None`` if not specified. While this device allows
+            for ``wires`` to be unspecified at construction time, other devices may make this argument
+            mandatory. Devices can also implement additional restrictions on the possible wires.
+        shots (int, Sequence[int], Sequence[Union[int, Sequence[int]]]): The default number of shots
+            to use in executions involving this device. Note that during execution, shots
+            are pulled from the circuit, not from the device.
+        seed (Union[str, None, int, array_like[int], SeedSequence, BitGenerator, Generator, jax.random.PRNGKey]): A
+            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng``. This is an optional
+            keyword argument added to follow recommend NumPy best practices. Other devices do not need
+            this parameter if it does not make sense for them.
+
+    """
+
+    name = "reference.qubit"
+
+    def __init__(self, wires=None, shots=None, seed=None):
+        super().__init__(wires=wires, shots=shots)
+
+        # seed and rng not necessary for a device, but part of recommended
+        # numpy practices to use a local random number generator
+        self._rng = np.random.default_rng(seed)
+
+    def preprocess(self, execution_config=DefaultExecutionConfig):
+
+        # Here we convert an arbitrary tape into one natively supported by the device
+        program = qml.transforms.core.TransformProgram()
+        program.add_transform(validate_device_wires, wires=self.wires, name="reference.qubit")
+        program.add_transform(qml.defer_measurements)
+        program.add_transform(qml.transforms.split_non_commuting)
+        program.add_transform(
+            decompose,
+            stopping_condition=supports_operation,
+            skip_initial_state_prep=False,
+            name="reference.qubit",
+        )
+        program.add_transform(qml.transforms.diagonalize_measurements)
```

Comment:
```suggestion
        program.add_transform(qml.transforms.diagonalize_measurements)
        program.add_transform(
            decompose,
            stopping_condition=supports_operation,
            skip_initial_state_prep=False,
            name="reference.qubit",
        )
```

Should we switch the order?